# Inspect 1886347_match.json

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
mpl.rcParams["animation.embed_limit"] = 50  # MB
import json

In [2]:
from google.colab import files

uploaded = files.upload()

Saving 1886347_match.json to 1886347_match.json


# match metadata

In [3]:
# Load the JSON file
with open("1886347_match.json", "r", encoding="utf-8") as f:
    match = json.load(f)

# See the top-level keys
print(match.keys())

dict_keys(['id', 'home_team_score', 'away_team_score', 'date_time', 'stadium', 'home_team', 'home_team_kit', 'away_team', 'away_team_kit', 'home_team_coach', 'away_team_coach', 'home_team_playing_time', 'away_team_playing_time', 'competition_edition', 'match_periods', 'competition_round', 'referees', 'players', 'status', 'home_team_side', 'ball', 'pitch_length', 'pitch_width'])


In [18]:
from pprint import pprint

pprint(match)

{'away_team': {'acronym': 'NEW',
               'id': 1805,
               'name': 'Newcastle United Jets FC',
               'short_name': 'Newcastle'},
 'away_team_coach': None,
 'away_team_kit': {'id': 10376,
                   'jersey_color': '#ffffff',
                   'name': 'away',
                   'number_color': '#000000',
                   'season': {'end_year': 2024,
                              'id': 29,
                              'name': '2024',
                              'start_year': 2024},
                   'team_id': 1805},
 'away_team_playing_time': {'minutes_otip': 31.13, 'minutes_tip': 22.2},
 'away_team_score': 0,
 'ball': {'trackable_object': 55},
 'competition_edition': {'competition': {'age_group': 'adult',
                                         'area': 'AUS',
                                         'gender': 'male',
                                         'id': 61,
                                         'name': 'A-League'},
                 

In [4]:
print(type(match["players"]))
print(len(match["players"]))

<class 'list'>
36


In [5]:
import pprint

pprint.pprint(match["players"][0])

{'birthday': '1998-03-11',
 'end_time': '01:25:21',
 'first_name': 'Guillermo Luis',
 'gender': 'male',
 'goal': 0,
 'id': 38673,
 'injured': False,
 'last_name': 'May Bartesaghi',
 'number': 10,
 'own_goal': 0,
 'player_role': {'acronym': 'CF',
                 'id': 15,
                 'name': 'Center Forward',
                 'position_group': 'Center Forward'},
 'playing_time': {'by_period': [{'end_frame': 27790,
                                 'minutes_otip': 11.03,
                                 'minutes_played': 46.3,
                                 'minutes_tip': 18.21,
                                 'name': 'period_1',
                                 'start_frame': 10},
                                {'end_frame': 52009,
                                 'minutes_otip': 7.73,
                                 'minutes_played': 40.35,
                                 'minutes_tip': 11.34,
                                 'name': 'period_2',
                             

| Field                                         | Meaning                                                                                                                              |
| --------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| `id`                                          | **Player ID** (38673). This identifies the player throughout the dataset.                                                            |
| `trackable_object`                            | **Tracking object ID** (39794). This is the key you'll use to link the player to the tracking data in `tracking_extrapolated.jsonl`. |
| `team_id`                                     | The team the player belongs to (4177).                                                                                               |
| `team_player_id`                              | An internal SkillCorner identifier.                                                                                                  |
| `first_name` / `last_name` / `short_name`     | Player names.                                                                                                                        |
| `number`                                      | Shirt number (10).                                                                                                                   |
| `birthday`                                    | Date of birth.                                                                                                                       |
| `player_role`                                 | Position information. Here it's **Center Forward (CF)**.                                                                             |
| `goal`, `own_goal`, `yellow_card`, `red_card` | Match statistics for the player.                                                                                                     |
| `start_time` / `end_time`                     | When the player was on the pitch.                                                                                                    |
| `playing_time`                                | Detailed playing time by half and total minutes.                                                                                     |


In [6]:
pprint.pprint(match["home_team"])
print()

pprint.pprint(match["away_team"])

{'acronym': 'AUC',
 'id': 4177,
 'name': 'Auckland FC',
 'short_name': 'Auckland FC'}

{'acronym': 'NEW',
 'id': 1805,
 'name': 'Newcastle United Jets FC',
 'short_name': 'Newcastle'}


In [20]:
# ============================================================
# GLOBAL VARIABLES
# ============================================================

home_team = match["home_team"]["name"]
away_team = match["away_team"]["name"]

home_team_id = match["home_team"]["id"]
away_team_id = match["away_team"]["id"]

pitch_length = match["pitch_length"]
pitch_width = match["pitch_width"]

FPS = 10

print(f"Home Team : {home_team}")
print(f"Away Team : {away_team}")
print(f"Pitch     : {pitch_length}m × {pitch_width}m")
print(f"FPS       : {FPS}")

Home Team : Auckland FC
Away Team : Newcastle United Jets FC
Pitch     : 104m × 68m
FPS       : 10


# Inspect 1886347_tracking_extrapolated.jsonl

In [7]:
from google.colab import files

uploaded = files.upload()

Saving 1886347_tracking_extrapolated.jsonl to 1886347_tracking_extrapolated.jsonl


In [8]:
import json

with open("1886347_tracking_extrapolated.jsonl", "r", encoding="utf-8") as f:
    first_frame = json.loads(next(f))

print(first_frame.keys())

dict_keys(['frame', 'timestamp', 'period', 'ball_data', 'possession', 'image_corners_projection', 'player_data'])


In [9]:
from pprint import pprint

pprint(first_frame)

{'ball_data': {'is_detected': None, 'x': None, 'y': None, 'z': None},
 'frame': 0,
 'image_corners_projection': {'x_bottom_left': None,
                              'x_bottom_right': None,
                              'x_top_left': None,
                              'x_top_right': None,
                              'y_bottom_left': None,
                              'y_bottom_right': None,
                              'y_top_left': None,
                              'y_top_right': None},
 'period': None,
 'player_data': [],
 'possession': {'group': None, 'player_id': None},
 'timestamp': None}


In [11]:
import json

with open("1886347_tracking_extrapolated.jsonl", "r", encoding="utf-8") as f:

    for i, line in enumerate(f):

        frame = json.loads(line)

        print(
            frame["frame"],
            frame["timestamp"]
        )

        if i == 9:
            break

0 None
1 None
2 None
3 None
4 None
5 None
6 None
7 None
8 None
9 None


In [13]:
import json

frames = []

with open(
    "1886347_tracking_extrapolated.jsonl",
    "r",
    encoding="utf-8"
) as f:

    for line in f:

        frames.append(json.loads(line))

print(f"Loaded {len(frames):,} frames")

Loaded 59,061 frames


# Match Possession Analysis

# Step 1: Create the function

In [22]:
# ============================================================
# MATCH POSSESSION ANALYSIS
# ============================================================

from collections import Counter

def calculate_match_possession(frames):

    possession_counts = Counter()

    # ---------------------------------------
    # Count possession frames
    # ---------------------------------------

    for frame in frames:

        possession = frame.get("possession", {})
        team = possession.get("group")

        if team == "home team":
            possession_counts["home"] += 1

        elif team == "away team":
            possession_counts["away"] += 1

        else:
            possession_counts["none"] += 1

    # ---------------------------------------
    # Frame totals
    # ---------------------------------------

    total_frames = (
        possession_counts["home"] +
        possession_counts["away"] +
        possession_counts["none"]
    )

    possession_frames = (
        possession_counts["home"] +
        possession_counts["away"]
    )

    # ---------------------------------------
    # Possession percentages
    # ---------------------------------------

    if possession_frames > 0:

        home_percent = (
            100 * possession_counts["home"] /
            possession_frames
        )

        away_percent = (
            100 * possession_counts["away"] /
            possession_frames
        )

    else:

        home_percent = 0
        away_percent = 0

    # ---------------------------------------
    # Tracking diagnostic
    # ---------------------------------------

    no_possession_percent = (
        100 * possession_counts["none"] /
        total_frames
    )

    # ---------------------------------------
    # Return results
    # ---------------------------------------

    return {

        "home_frames": possession_counts["home"],
        "away_frames": possession_counts["away"],
        "no_possession_frames": possession_counts["none"],

        "home_percent": home_percent,
        "away_percent": away_percent,

        "no_possession_percent": no_possession_percent,

        "total_frames": total_frames,
        "possession_frames": possession_frames
    }

In [25]:
possession = calculate_match_possession(frames)

print("=" * 60)
print("MATCH POSSESSION ANALYSIS")
print("=" * 60)

print("\nOfficial Possession")
print(f"{home_team:<30} {possession['home_percent']:.2f}%")
print(f"{away_team:<30} {possession['away_percent']:.2f}%")

print("\nTracking Diagnostics")
print(f"Home Frames              : {possession['home_frames']:,}")
print(f"Away Frames              : {possession['away_frames']:,}")
print(f"No Possession Frames     : {possession['no_possession_frames']:,}")
print(f"Total Frames             : {possession['total_frames']:,}")
print(f"No Possession Rate       : {possession['no_possession_percent']:.2f}%")

MATCH POSSESSION ANALYSIS

Official Possession
Auckland FC                    57.45%
Newcastle United Jets FC       42.55%

Tracking Diagnostics
Home Frames              : 18,441
Away Frames              : 13,659
No Possession Frames     : 26,961
Total Frames             : 59,061
No Possession Rate       : 45.65%


# Distance Covered


# Create the function

In [26]:
# ============================================================
# MATCH DISTANCE COVERED
# ============================================================

import math
from collections import defaultdict

def calculate_match_distance(frames):

    # Total distance covered by each player (metres)
    distance = defaultdict(float)

    # Previous position of each player
    previous_position = {}

    for frame in frames:

        for player in frame["player_data"]:

            # Ignore extrapolated positions
            if not player["is_detected"]:
                continue

            player_id = player["player_id"]

            x = player["x"]
            y = player["y"]

            # First observation
            if player_id not in previous_position:

                previous_position[player_id] = (x, y)
                continue

            prev_x, prev_y = previous_position[player_id]

            # Euclidean distance
            d = math.sqrt(
                (x - prev_x) ** 2 +
                (y - prev_y) ** 2
            )

            distance[player_id] += d

            previous_position[player_id] = (x, y)

    return dict(distance)

In [34]:
print(type(match["players"]))
print(len(match["players"]))

<class 'list'>
36


In [35]:
# ============================================================
# PLAYER METADATA
# ============================================================

players = match["players"]

print(f"Loaded {len(players)} players")

Loaded 36 players


In [36]:
# ============================================================
# PLAYER LOOKUP
# ============================================================

player_lookup = {}

for player in players:

    player_lookup[player["id"]] = player

print(f"Player lookup created for {len(player_lookup)} players.")

Player lookup created for 36 players.


In [29]:
distance = calculate_match_distance(frames)

In [37]:
print("=" * 60)
print("DISTANCE COVERED")
print("=" * 60)

for player_id, metres in sorted(
    distance.items(),
    key=lambda x: x[1],
    reverse=True
):

    name = player_lookup[player_id]["short_name"]

    print(
        f"{name:<25}"
        f"{metres/1000:.2f} km"
    )

DISTANCE COVERED
F. Gallegos              11.90 km
L. Verstraete            11.16 km
T. Aquilina              10.79 km
L. Gillion               10.43 km
K. Grozos                10.23 km
C. Timmins               10.11 km
F. De Vries              9.63 km
E. Adams                 9.24 km
D. Ingham                9.08 km
C. Elliott               9.00 km
M. Natta                 8.78 km
J. Brimmer               8.37 km
L. Rogerson              8.25 km
A. Šušnjar               7.92 km
G. May                   7.72 km
P. Cancar                7.67 km
L. Bayliss               7.44 km
N. Pijnaker              7.08 km
B. Gibson                7.04 km
D. Hall                  7.02 km
D. Wilmering             3.32 km
N. Moreno                2.68 km
C. Taylor                2.61 km
M. Mata                  1.90 km
R. Scott                 1.44 km
A. Paulsen               1.23 km
M. Scarcella             1.14 km
J. Vidic                 1.02 km
T. Smith                 0.91 km


In [38]:
home_total = 0
away_total = 0

for player_id, metres in distance.items():

    info = player_lookup[player_id]

    if info["team_id"] == home_team_id:
        home_total += metres
    else:
        away_total += metres

print("\n")
print("=" * 60)
print("TEAM DISTANCE")
print("=" * 60)

print(f"{home_team:<30}{home_total/1000:.2f} km")
print(f"{away_team:<30}{away_total/1000:.2f} km")



TEAM DISTANCE
Auckland FC                   97.29 km
Newcastle United Jets FC      97.80 km


In [40]:
# ============================================================
# TEAM DISTANCE COVERED
# ============================================================

home_distance = 0
away_distance = 0

for player_id, metres in distance.items():

    info = player_lookup[player_id]

    if info["team_id"] == home_team_id:
        home_distance += metres
    else:
        away_distance += metres

difference = abs(home_distance - away_distance)

print("\n" + "=" * 60)
print("TEAM DISTANCE")
print("=" * 60)

print(f"{home_team:<30}{home_distance/1000:.2f} km")
print(f"{away_team:<30}{away_distance/1000:.2f} km")

print("-" * 60)

print(f"Difference{'':<20}{difference/1000:.2f} km")


TEAM DISTANCE
Auckland FC                   97.29 km
Newcastle United Jets FC      97.80 km
------------------------------------------------------------
Difference                    0.51 km


# Sprint Analysis

# Create the function

In [56]:
# ============================================================
# MATCH SPRINT ANALYSIS
# ============================================================

import math
from collections import defaultdict

FPS = 10
SPRINT_THRESHOLD = 25.0      # km/h
MAX_VALID_SPEED = 38.5       # km/h
MIN_MOVEMENT = 0.05          # metres


def calculate_sprints(frames):

    sprint_count = defaultdict(int)
    max_speed = defaultdict(float)

    # Store previous x, y and frame number
    previous_position = {}

    # Track whether a player is currently sprinting
    is_sprinting = defaultdict(bool)

    for frame in frames:

        current_frame = frame["frame"]

        for player in frame["player_data"]:

            # Ignore extrapolated positions
            if not player["is_detected"]:
                continue

            player_id = player["player_id"]

            x = player["x"]
            y = player["y"]

            # First observation of this player
            if player_id not in previous_position:

                previous_position[player_id] = (
                    x,
                    y,
                    current_frame
                )

                continue

            # Previous position
            prev_x, prev_y, prev_frame = previous_position[player_id]

            # ----------------------------------------------------
            # Distance travelled (metres)
            # ----------------------------------------------------

            distance = math.sqrt(
                (x - prev_x) ** 2 +
                (y - prev_y) ** 2
            )

            # Ignore tiny tracking jitter
            if distance < MIN_MOVEMENT:
                distance = 0

            # ----------------------------------------------------
            # Time elapsed
            # ----------------------------------------------------

            frame_gap = current_frame - prev_frame

            if frame_gap <= 0:

                previous_position[player_id] = (
                    x,
                    y,
                    current_frame
                )

                continue

            time_elapsed = frame_gap / FPS

            # ----------------------------------------------------
            # Speed (km/h)
            # ----------------------------------------------------

            speed = (distance / time_elapsed) * 3.6

            # Ignore impossible tracking jumps
            if speed > MAX_VALID_SPEED:

                previous_position[player_id] = (
                    x,
                    y,
                    current_frame
                )

                continue

            # Record maximum speed
            if speed > max_speed[player_id]:
                max_speed[player_id] = speed

            # ----------------------------------------------------
            # Sprint Detection
            # ----------------------------------------------------

            if speed >= SPRINT_THRESHOLD:

                # Count only when a sprint starts
                if not is_sprinting[player_id]:

                    sprint_count[player_id] += 1
                    is_sprinting[player_id] = True

            else:

                # Sprint has ended
                is_sprinting[player_id] = False

            # Update previous position
            previous_position[player_id] = (
                x,
                y,
                current_frame
            )

    return sprint_count, max_speed

In [57]:
sprint_count, max_speed = calculate_sprints(frames)

# Sprint leaderboard

In [58]:
print("=" * 60)
print("SPRINT LEADERBOARD")
print("=" * 60)

for player_id, sprints in sorted(
    sprint_count.items(),
    key=lambda x: x[1],
    reverse=True
):

    name = player_lookup[player_id]["short_name"]

    print(
        f"{name:<25}"
        f"{sprints:>5} sprints"
    )

SPRINT LEADERBOARD
L. Gillion                  27 sprints
T. Aquilina                 25 sprints
D. Ingham                   23 sprints
F. Gallegos                 22 sprints
M. Natta                    20 sprints
C. Elliott                  20 sprints
G. May                      19 sprints
E. Adams                    18 sprints
J. Brimmer                  18 sprints
L. Rogerson                 18 sprints
P. Cancar                   16 sprints
F. De Vries                 14 sprints
B. Gibson                   13 sprints
C. Timmins                  13 sprints
L. Verstraete               12 sprints
L. Bayliss                  11 sprints
D. Hall                     10 sprints
K. Grozos                   10 sprints
C. Taylor                   10 sprints
N. Pijnaker                  9 sprints
D. Wilmering                 9 sprints
N. Moreno                    9 sprints
A. Šušnjar                   8 sprints
M. Mata                      4 sprints
J. Vidic                     2 sprints
T. Smi

# Fastest players

In [59]:
print("\n" + "=" * 60)
print("FASTEST PLAYERS")
print("=" * 60)

for player_id, speed in sorted(
    max_speed.items(),
    key=lambda x: x[1],
    reverse=True
):

    name = player_lookup[player_id]["short_name"]

    print(
        f"{name:<25}"
        f"{speed:.1f} km/h"
    )


FASTEST PLAYERS
J. Brimmer               38.1 km/h
C. Elliott               37.8 km/h
L. Gillion               37.7 km/h
J. Vidic                 36.7 km/h
D. Hall                  35.4 km/h
G. May                   33.7 km/h
N. Pijnaker              33.5 km/h
P. Cancar                33.3 km/h
E. Adams                 32.2 km/h
M. Natta                 32.1 km/h
C. Timmins               31.8 km/h
F. Gallegos              31.7 km/h
N. Moreno                31.5 km/h
L. Rogerson              31.5 km/h
D. Wilmering             31.5 km/h
T. Smith                 31.5 km/h
T. Aquilina              31.4 km/h
A. Šušnjar               31.2 km/h
B. Gibson                31.0 km/h
K. Grozos                30.8 km/h
M. Mata                  30.7 km/h
D. Ingham                30.4 km/h
C. Taylor                30.0 km/h
L. Verstraete            29.8 km/h
A. Paulsen               29.7 km/h
F. De Vries              29.6 km/h
L. Bayliss               27.9 km/h
M. Scarcella             23.3 km/h
R. 

# Team sprint totals

In [61]:
home_sprints = 0
away_sprints = 0

for player_id, sprints in sprint_count.items():

    info = player_lookup[player_id]

    if info["team_id"] == home_team_id:
        home_sprints += sprints
    else:
        away_sprints += sprints

print("\n" + "=" * 60)
print("TEAM SPRINTS")
print("=" * 60)

print(f"{home_team:<30}{home_sprints}")
print(f"{away_team:<30}{away_sprints}")


TEAM SPRINTS
Auckland FC                   184
Newcastle United Jets FC      178


# Top Speed

# Top Speed Leaderboard

In [62]:
# ============================================================
# TOP SPEED LEADERBOARD
# ============================================================

print("=" * 60)
print("TOP SPEED LEADERBOARD")
print("=" * 60)

for rank, (player_id, speed) in enumerate(
    sorted(
        max_speed.items(),
        key=lambda x: x[1],
        reverse=True
    ),
    start=1
):

    name = player_lookup[player_id]["short_name"]

    print(
        f"{rank:>2}. "
        f"{name:<25}"
        f"{speed:>6.1f} km/h"
    )

TOP SPEED LEADERBOARD
 1. J. Brimmer                 38.1 km/h
 2. C. Elliott                 37.8 km/h
 3. L. Gillion                 37.7 km/h
 4. J. Vidic                   36.7 km/h
 5. D. Hall                    35.4 km/h
 6. G. May                     33.7 km/h
 7. N. Pijnaker                33.5 km/h
 8. P. Cancar                  33.3 km/h
 9. E. Adams                   32.2 km/h
10. M. Natta                   32.1 km/h
11. C. Timmins                 31.8 km/h
12. F. Gallegos                31.7 km/h
13. N. Moreno                  31.5 km/h
14. L. Rogerson                31.5 km/h
15. D. Wilmering               31.5 km/h
16. T. Smith                   31.5 km/h
17. T. Aquilina                31.4 km/h
18. A. Šušnjar                 31.2 km/h
19. B. Gibson                  31.0 km/h
20. K. Grozos                  30.8 km/h
21. M. Mata                    30.7 km/h
22. D. Ingham                  30.4 km/h
23. C. Taylor                  30.0 km/h
24. L. Verstraete              29.8

In [65]:
print("=" * 60)
print("TOP SPEED LEADERBOARD")
print("=" * 60)

for rank, (player_id, speed) in enumerate(
    sorted(
        max_speed.items(),
        key=lambda x: x[1],
        reverse=True
    )[:10],
    start=1
):

    name = player_lookup[player_id]["short_name"]

    print(
        f"{rank:>2}. "
        f"{name:<25}"
        f"{speed:>6.1f} km/h"
    )

TOP SPEED LEADERBOARD
 1. J. Brimmer                 38.1 km/h
 2. C. Elliott                 37.8 km/h
 3. L. Gillion                 37.7 km/h
 4. J. Vidic                   36.7 km/h
 5. D. Hall                    35.4 km/h
 6. G. May                     33.7 km/h
 7. N. Pijnaker                33.5 km/h
 8. P. Cancar                  33.3 km/h
 9. E. Adams                   32.2 km/h
10. M. Natta                   32.1 km/h


# Team Top Speed

In [63]:
# ============================================================
# TEAM TOP SPEED
# ============================================================

home_fastest = ("", 0)
away_fastest = ("", 0)

for player_id, speed in max_speed.items():

    info = player_lookup[player_id]
    name = info["short_name"]

    if info["team_id"] == home_team_id:

        if speed > home_fastest[1]:
            home_fastest = (name, speed)

    else:

        if speed > away_fastest[1]:
            away_fastest = (name, speed)

print("\n" + "=" * 60)
print("TEAM TOP SPEED")
print("=" * 60)

print(
    f"{home_team:<30}"
    f"{home_fastest[0]} "
    f"({home_fastest[1]:.1f} km/h)"
)

print(
    f"{away_team:<30}"
    f"{away_fastest[0]} "
    f"({away_fastest[1]:.1f} km/h)"
)


TEAM TOP SPEED
Auckland FC                   J. Brimmer (38.1 km/h)
Newcastle United Jets FC      J. Vidic (36.7 km/h)


# Match Top Speed

In [64]:
# ============================================================
# MATCH TOP SPEED
# ============================================================

fastest_player = max(
    max_speed.items(),
    key=lambda x: x[1]
)

player_id = fastest_player[0]
speed = fastest_player[1]

name = player_lookup[player_id]["short_name"]

print("\n" + "=" * 60)
print("MATCH TOP SPEED")
print("=" * 60)

print(f"Player : {name}")
print(f"Speed  : {speed:.1f} km/h")


MATCH TOP SPEED
Player : J. Brimmer
Speed  : 38.1 km/h


In [131]:
# ============================================================
# MATCH TOP SPEED
# ============================================================

if home_fastest[1] > away_fastest[1]:
    fastest_player = home_fastest[0]
    fastest_speed = home_fastest[1]
else:
    fastest_player = away_fastest[0]
    fastest_speed = away_fastest[1]

In [132]:
print(f"{'Fastest Player'}")
print(f"{fastest_player:<35}{fastest_speed:.1f} km/h")

Fastest Player
J. Brimmer                         38.1 km/h


# Team Width Analysis
Team Width = Right-most player − Left-most player

# Create the Team Width Function

In [70]:
# ============================================================
# TEAM WIDTH ANALYSIS
# ============================================================

import numpy as np

def calculate_team_width(frames, team_id):

    widths = []

    for frame in frames:

        y_positions = []

        for player in frame["player_data"]:

            # Ignore undetected players
            if not player["is_detected"]:
                continue

            info = player_lookup.get(player["player_id"])

            if info is None:
                continue

            if info["team_id"] != team_id:
                continue

            # Use x-coordinate (pitch length direction)
            y_positions.append(player["y"])

        # Need enough players to calculate width
        if len(y_positions) < 4:
            continue

        width = max(y_positions) - min(y_positions)

        widths.append(width)

    return widths

# Calculate Both Teams

In [71]:
home_widths = calculate_team_width(frames, home_team_id)
away_widths = calculate_team_width(frames, away_team_id)

# Match Summary

In [73]:
# ============================================================
# TEAM WIDTH SUMMARY
# ============================================================

print("=" * 60)
print("TEAM WIDTH")
print("=" * 60)

print(f"{home_team}")
print(f"Average : {np.mean(home_widths):.2f} m")
print(f"Minimum : {np.min(home_widths):.2f} m")
print(f"Maximum : {np.max(home_widths):.2f} m")
print(f"Std Dev : {np.std(home_widths):.2f} m")

print()

print(f"{away_team}")
print(f"Average : {np.mean(away_widths):.2f} m")
print(f"Minimum : {np.min(away_widths):.2f} m")
print(f"Maximum : {np.max(away_widths):.2f} m")
print(f"Std Dev : {np.std(away_widths):.2f} m")

TEAM WIDTH
Auckland FC
Average : 37.27 m
Minimum : 7.10 m
Maximum : 64.15 m
Std Dev : 10.87 m

Newcastle United Jets FC
Average : 36.28 m
Minimum : 3.64 m
Maximum : 66.49 m
Std Dev : 10.53 m


# Compare the Teams

In [74]:
# ============================================================
# TEAM WIDTH COMPARISON
# ============================================================

home_avg = np.mean(home_widths)
away_avg = np.mean(away_widths)

difference = abs(home_avg - away_avg)

print("\n" + "=" * 60)
print("TEAM WIDTH COMPARISON")
print("=" * 60)

print(f"{home_team:<30}{home_avg:.2f} m")
print(f"{away_team:<30}{away_avg:.2f} m")

print("-" * 60)
print(f"Difference{'':<20}{difference:.2f} m")


TEAM WIDTH COMPARISON
Auckland FC                   37.27 m
Newcastle United Jets FC      36.28 m
------------------------------------------------------------
Difference                    0.99 m


# Team Length Analysis
How stretched is the team from the deepest player to the highest player?

# Calculate Team Length

In [91]:
# ============================================================
# TEAM LENGTH ANALYSIS
# ============================================================

import numpy as np

MIN_PLAYERS = 9      # Minimum detected players required

def calculate_team_length(frames, team_id):

    lengths = []

    for frame in frames:

        x_positions = []

        for player in frame["player_data"]:

            # Ignore undetected players
            if not player["is_detected"]:
                continue

            info = player_lookup.get(player["player_id"])

            if info is None:
                continue

            # Keep only this team's players
            if info["team_id"] != team_id:
                continue

            x_positions.append(player["x"])

        # Ignore poor tracking frames
        if len(x_positions) < MIN_PLAYERS:
            continue

        team_length = max(x_positions) - min(x_positions)

        lengths.append(team_length)

    return lengths

# Calculate Both Teams

In [94]:
home_lengths = calculate_team_length(frames, home_team_id)
away_lengths = calculate_team_length(frames, away_team_id)

In [95]:
print(home_lengths[:10])
print()
print(home_widths[:10])

[17.450000000000003, 17.48, 17.5, 17.529999999999998, 17.549999999999997, 17.57, 17.59, 17.610000000000003, 17.62, 17.62]

[30.96, 30.97, 30.98, 30.990000000000002, 31.0, 31.009999999999998, 31.020000000000003, 31.029999999999998, 31.04, 31.05]


# Team Length Summary

In [96]:
# ============================================================
# TEAM LENGTH SUMMARY
# ============================================================

print("=" * 60)
print("TEAM LENGTH")
print("=" * 60)

print(f"{home_team}")
print(f"Average : {np.mean(home_lengths):.2f} m")
print(f"Minimum : {np.min(home_lengths):.2f} m")
print(f"Maximum : {np.max(home_lengths):.2f} m")
print(f"Std Dev : {np.std(home_lengths):.2f} m")

print()

print(f"{away_team}")
print(f"Average : {np.mean(away_lengths):.2f} m")
print(f"Minimum : {np.min(away_lengths):.2f} m")
print(f"Maximum : {np.max(away_lengths):.2f} m")
print(f"Std Dev : {np.std(away_lengths):.2f} m")

TEAM LENGTH
Auckland FC
Average : 24.15 m
Minimum : 8.03 m
Maximum : 45.50 m
Std Dev : 6.08 m

Newcastle United Jets FC
Average : 22.14 m
Minimum : 5.05 m
Maximum : 41.17 m
Std Dev : 6.53 m


In [97]:
frame = frames[0]

for player in frame["player_data"]:

    if player["is_detected"]:
        print(player["x"], player["y"])

In [98]:
# ============================================================
# CHECK X/Y COORDINATE RANGES
# ============================================================

for frame in frames:

    x_values = []
    y_values = []

    for player in frame["player_data"]:

        if player["is_detected"]:
            x_values.append(player["x"])
            y_values.append(player["y"])

    if len(x_values) > 0:

        print("Frame:", frame["frame"])
        print("Timestamp:", frame["timestamp"])
        print("X Range:", min(x_values), "to", max(x_values))
        print("Y Range:", min(y_values), "to", max(y_values))

        break

Frame: 10
Timestamp: 00:00:00.00
X Range: -21.83 to 17.85
Y Range: -32.47 to 18.8


In [99]:
print(f"Average : {np.mean(home_lengths):.2f} m")
print(f"Median  : {np.median(home_lengths):.2f} m")
print(f"5th %   : {np.percentile(home_lengths, 5):.2f} m")
print(f"95th %  : {np.percentile(home_lengths, 95):.2f} m")

Average : 24.15 m
Median  : 24.77 m
5th %   : 13.13 m
95th %  : 32.84 m


In [100]:
max_length = max(home_lengths)
max_index = home_lengths.index(max_length)

print("Maximum Length:", max_length)
print("Frame Index:", max_index)
print("Actual Frame Number:", frames[max_index]["frame"])

Maximum Length: 45.5
Frame Index: 8563
Actual Frame Number: 8563


In [101]:
frame = frames[max_index]

x_positions = []

for player in frame["player_data"]:

    if not player["is_detected"]:
        continue

    info = player_lookup.get(player["player_id"])

    if info and info["team_id"] == home_team_id:
        x_positions.append(player["x"])

print(sorted(x_positions))

[-31.42, -30.61, -24.88, -24.67, -17.16, -15.24, -15.14, -13.65]


# Team Length Comparison

In [102]:
# ============================================================
# TEAM LENGTH COMPARISON
# ============================================================

home_avg = np.mean(home_lengths)
away_avg = np.mean(away_lengths)

difference = abs(home_avg - away_avg)

print("\n" + "=" * 60)
print("TEAM LENGTH COMPARISON")
print("=" * 60)

print(f"{home_team:<30}{home_avg:.2f} m")
print(f"{away_team:<30}{away_avg:.2f} m")

print("-" * 60)
print(f"Difference{'':<20}{difference:.2f} m")


TEAM LENGTH COMPARISON
Auckland FC                   24.15 m
Newcastle United Jets FC      22.14 m
------------------------------------------------------------
Difference                    2.01 m


# Convex Hull Area
How much space does the team occupy?

# Import ConvexHull

In [103]:
from scipy.spatial import ConvexHull
import numpy as np

# Calculate Convex Hull Area

In [104]:
# ============================================================
# CONVEX HULL AREA
# ============================================================

from scipy.spatial import ConvexHull

MIN_PLAYERS = 9

def calculate_convex_hull_area(frames, team_id):

    areas = []

    for frame in frames:

        points = []

        for player in frame["player_data"]:

            # Ignore undetected players
            if not player["is_detected"]:
                continue

            info = player_lookup.get(player["player_id"])

            if info is None:
                continue

            if info["team_id"] != team_id:
                continue

            points.append([
                player["x"],
                player["y"]
            ])

        # Require enough detected players
        if len(points) < MIN_PLAYERS:
            continue

        points = np.array(points)

        hull = ConvexHull(points)

        areas.append(hull.volume)

    return areas

# Calculate Both Teams

In [105]:
home_hull = calculate_convex_hull_area(
    frames,
    home_team_id
)

away_hull = calculate_convex_hull_area(
    frames,
    away_team_id
)

# Match Summary

In [106]:
# ============================================================
# CONVEX HULL SUMMARY
# ============================================================

print("=" * 60)
print("CONVEX HULL AREA")
print("=" * 60)

print(home_team)
print(f"Average : {np.mean(home_hull):.1f} m²")
print(f"Minimum : {np.min(home_hull):.1f} m²")
print(f"Maximum : {np.max(home_hull):.1f} m²")
print(f"Std Dev : {np.std(home_hull):.1f} m²")

print()

print(away_team)
print(f"Average : {np.mean(away_hull):.1f} m²")
print(f"Minimum : {np.min(away_hull):.1f} m²")
print(f"Maximum : {np.max(away_hull):.1f} m²")
print(f"Std Dev : {np.std(away_hull):.1f} m²")

CONVEX HULL AREA
Auckland FC
Average : 595.3 m²
Minimum : 65.0 m²
Maximum : 1679.5 m²
Std Dev : 289.0 m²

Newcastle United Jets FC
Average : 513.9 m²
Minimum : 32.1 m²
Maximum : 1536.6 m²
Std Dev : 268.2 m²


# Team Comparison

In [107]:
home_avg = np.mean(home_hull)
away_avg = np.mean(away_hull)

difference = abs(home_avg - away_avg)

print("\n" + "=" * 60)
print("CONVEX HULL COMPARISON")
print("=" * 60)

print(f"{home_team:<30}{home_avg:.1f} m²")
print(f"{away_team:<30}{away_avg:.1f} m²")

print("-" * 60)
print(f"Difference{'':<20}{difference:.1f} m²")


CONVEX HULL COMPARISON
Auckland FC                   595.3 m²
Newcastle United Jets FC      513.9 m²
------------------------------------------------------------
Difference                    81.4 m²


# Team Compactness
Team compactness measures how closely connected a team's players are.

# Calculate Compactness

In [116]:
# ============================================================
# TEAM SHAPE ANALYSIS
# Width + Length + Convex Hull + Compactness
# ============================================================

import numpy as np
from scipy.spatial import ConvexHull

MIN_PLAYERS = 9

def calculate_team_shape(frames, team_id):

    results = []

    for frame in frames:

        points = []

        for player in frame["player_data"]:

            if not player["is_detected"]:
                continue

            info = player_lookup.get(player["player_id"])

            if info is None:
                continue

            if info["team_id"] != team_id:
                continue

            points.append((player["x"], player["y"]))

        # Ignore poor tracking frames
        if len(points) < MIN_PLAYERS:
            continue

        points = np.array(points)

        x = points[:, 0]
        y = points[:, 1]

        # -------------------------
        # Width & Length
        # -------------------------

        width = np.max(y) - np.min(y)
        length = np.max(x) - np.min(x)

        # -------------------------
        # Convex Hull
        # -------------------------

        hull = ConvexHull(points)
        hull_area = hull.volume

        # -------------------------
        # Compactness
        # -------------------------

        rectangle_area = width * length

        if rectangle_area == 0:
            continue

        compactness = hull_area / rectangle_area

        results.append({

            "frame": frame["frame"],
            "timestamp": frame["timestamp"],

            "width": width,
            "length": length,

            "hull_area": hull_area,
            "compactness": compactness

        })

    return results

# Calculate Both Teams

In [118]:
home_shape = calculate_team_shape(
    frames,
    home_team_id
)

away_shape = calculate_team_shape(
    frames,
    away_team_id
)

# Extract the metrics

In [120]:
home_widths = [x["width"] for x in home_shape]
home_lengths = [x["length"] for x in home_shape]
home_hull = [x["hull_area"] for x in home_shape]
home_compactness = [x["compactness"] for x in home_shape]

away_widths = [x["width"] for x in away_shape]
away_lengths = [x["length"] for x in away_shape]
away_hull = [x["hull_area"] for x in away_shape]
away_compactness = [x["compactness"] for x in away_shape]

# Summary

In [122]:
# ============================================================
# TEAM COMPACTNESS SUMMARY
# ============================================================

print("=" * 60)
print("TEAM COMPACTNESS")
print("=" * 60)

print(home_team)
print(f"Average : {np.mean(home_compactness):.3f}")
print(f"Minimum : {np.min(home_compactness):.3f}")
print(f"Maximum : {np.max(home_compactness):.3f}")
print(f"Std Dev : {np.std(home_compactness):.3f}")

print()

print(away_team)
print(f"Average : {np.mean(away_compactness):.3f}")
print(f"Minimum : {np.min(away_compactness):.3f}")
print(f"Maximum : {np.max(away_compactness):.3f}")
print(f"Std Dev : {np.std(away_compactness):.3f}")

TEAM COMPACTNESS
Auckland FC
Average : 0.640
Minimum : 0.188
Maximum : 0.895
Std Dev : 0.093

Newcastle United Jets FC
Average : 0.634
Minimum : 0.184
Maximum : 0.929
Std Dev : 0.100


# Comparison

In [125]:
home_avg = np.mean(home_compactness)
away_avg = np.mean(away_compactness)

print("\n" + "=" * 60)
print("TEAM COMPACTNESS COMPARISON")
print("=" * 60)

print(f"{home_team:<30}{home_avg:.3f}")
print(f"{away_team:<30}{away_avg:.3f}")

print("-" * 60)
print(f"Difference{'':<20}{abs(home_avg-away_avg):.3f}")


TEAM COMPACTNESS COMPARISON
Auckland FC                   0.640
Newcastle United Jets FC      0.634
------------------------------------------------------------
Difference                    0.006


# Space Control
How much of the pitch each team controls at any moment?




# Create the Pitch Grid

In [126]:
import numpy as np

GRID_SIZE = 2

x_grid = np.arange(-52.5, 52.5, GRID_SIZE)
y_grid = np.arange(-34, 34, GRID_SIZE)

# Space Control Function

In [127]:
# ============================================================
# SPACE CONTROL
# ============================================================

import numpy as np

MIN_PLAYERS = 9

def calculate_space_control(frames):

    results = []

    for frame in frames:

        home_players = []
        away_players = []

        for player in frame["player_data"]:

            if not player["is_detected"]:
                continue

            info = player_lookup.get(player["player_id"])

            if info is None:
                continue

            point = np.array([player["x"], player["y"]])

            if info["team_id"] == home_team_id:
                home_players.append(point)
            else:
                away_players.append(point)

        if len(home_players) < MIN_PLAYERS:
            continue

        if len(away_players) < MIN_PLAYERS:
            continue

        home_cells = 0
        away_cells = 0

        for x in x_grid:
            for y in y_grid:

                location = np.array([x, y])

                home_distance = min(
                    np.linalg.norm(location - p)
                    for p in home_players
                )

                away_distance = min(
                    np.linalg.norm(location - p)
                    for p in away_players
                )

                if home_distance < away_distance:
                    home_cells += 1
                else:
                    away_cells += 1

        total = home_cells + away_cells

        results.append({

            "frame": frame["frame"],
            "timestamp": frame["timestamp"],

            "home_space": 100 * home_cells / total,
            "away_space": 100 * away_cells / total

        })

    return results

In [128]:
space_control = calculate_space_control(frames)

In [129]:
home_space = [x["home_space"] for x in space_control]
away_space = [x["away_space"] for x in space_control]

print("=" * 60)
print("SPACE CONTROL")
print("=" * 60)

print(f"{home_team:<30}{np.mean(home_space):.2f}%")
print(f"{away_team:<30}{np.mean(away_space):.2f}%")

SPACE CONTROL
Auckland FC                   57.34%
Newcastle United Jets FC      42.66%


In [133]:
# ============================================================
# MATCH DASHBOARD
# ============================================================

print("=" * 70)
print("MATCH DASHBOARD".center(70))
print("=" * 70)

print(f"\nMatch")
print("-" * 70)
print(f"{home_team} vs {away_team}")

print("\nPOSSESSION")
print("-" * 70)
print(f"{home_team:<35}{possession['home_percent']:.2f}%")
print(f"{away_team:<35}{possession['away_percent']:.2f}%")

print("\nSPACE CONTROL")
print("-" * 70)
print(f"{home_team:<35}{np.mean(home_space):.2f}%")
print(f"{away_team:<35}{np.mean(away_space):.2f}%")

print("\nPHYSICAL PERFORMANCE")
print("-" * 70)
print(f"{'Distance Covered'}")
print(f"{home_team:<35}{home_distance:.2f} km")
print(f"{away_team:<35}{away_distance:.2f} km")

print()

print(f"{'Team Sprints'}")
print(f"{home_team:<35}{home_sprints}")
print(f"{away_team:<35}{away_sprints}")

print()

print(f"{'Fastest Player'}")
print(f"{fastest_player:<35}{fastest_speed:.1f} km/h")

print("\nTACTICAL SHAPE")
print("-" * 70)

print(f"{'Width'}")
print(f"{home_team:<35}{np.mean(home_widths):.2f} m")
print(f"{away_team:<35}{np.mean(away_widths):.2f} m")

print()

print(f"{'Length'}")
print(f"{home_team:<35}{np.mean(home_lengths):.2f} m")
print(f"{away_team:<35}{np.mean(away_lengths):.2f} m")

print()

print(f"{'Convex Hull Area'}")
print(f"{home_team:<35}{np.mean(home_hull):.1f} m²")
print(f"{away_team:<35}{np.mean(away_hull):.1f} m²")

print()

print(f"{'Compactness'}")
print(f"{home_team:<35}{np.mean(home_compactness):.3f}")
print(f"{away_team:<35}{np.mean(away_compactness):.3f}")

print("\n" + "=" * 70)

                           MATCH DASHBOARD                            

Match
----------------------------------------------------------------------
Auckland FC vs Newcastle United Jets FC

POSSESSION
----------------------------------------------------------------------
Auckland FC                        57.45%
Newcastle United Jets FC           42.55%

SPACE CONTROL
----------------------------------------------------------------------
Auckland FC                        57.34%
Newcastle United Jets FC           42.66%

PHYSICAL PERFORMANCE
----------------------------------------------------------------------
Distance Covered
Auckland FC                        97290.69 km
Newcastle United Jets FC           97804.17 km

Team Sprints
Auckland FC                        184
Newcastle United Jets FC           178

Fastest Player
J. Brimmer                         38.1 km/h

TACTICAL SHAPE
----------------------------------------------------------------------
Width
Auckland FC             

# Automatic Match Report

In [134]:
# ============================================================
# AUTOMATIC MATCH REPORT
# ============================================================

print("=" * 70)
print("AUTOMATIC MATCH REPORT".center(70))
print("=" * 70)

print(
    f"\n{home_team} controlled the match with "
    f"{possession['home_percent']:.2f}% possession "
    f"compared to {away_team}'s "
    f"{possession['away_percent']:.2f}%."
)

print(
    f"\nTerritorially, {home_team} controlled "
    f"{np.mean(home_space):.2f}% of the pitch, "
    f"while {away_team} controlled "
    f"{np.mean(away_space):.2f}%."
)

print(
    f"\nBoth teams produced a similar physical output."
)

print(
    f"{home_team} covered "
    f"{home_distance:.2f} km, "
    f"while {away_team} covered "
    f"{away_distance:.2f} km."
)

print(
    f"\nSprint totals were "
    f"{home_sprints} for {home_team} "
    f"and {away_sprints} for {away_team}."
)

if home_fastest[1] > away_fastest[1]:

    fastest_name = home_fastest[0]
    fastest_speed = home_fastest[1]

else:

    fastest_name = away_fastest[0]
    fastest_speed = away_fastest[1]

print(
    f"\nThe fastest player was "
    f"{fastest_name} "
    f"who reached "
    f"{fastest_speed:.1f} km/h."
)

print(
    f"\nTactically, {home_team} maintained a slightly "
    f"wider shape "
    f"({np.mean(home_widths):.2f} m vs "
    f"{np.mean(away_widths):.2f} m)"
)

print(
    f"and a longer team structure "
    f"({np.mean(home_lengths):.2f} m vs "
    f"{np.mean(away_lengths):.2f} m)."
)

print(
    f"\nThe average Convex Hull Area was "
    f"{np.mean(home_hull):.1f} m² for {home_team} "
    f"and {np.mean(away_hull):.1f} m² for "
    f"{away_team}, indicating that "
    f"{home_team} occupied more space throughout the match."
)

print(
    f"\nBoth teams maintained very similar compactness "
    f"({np.mean(home_compactness):.3f} vs "
    f"{np.mean(away_compactness):.3f}), "
    f"suggesting comparable defensive organization."
)

print(
    f"\nOverall, the tracking data suggests that "
    f"{home_team} controlled possession, territory, "
    f"and team shape, while {away_team} remained "
    f"competitive through similar physical output "
    f"and defensive compactness."
)

print("\n" + "=" * 70)

                        AUTOMATIC MATCH REPORT                        

Auckland FC controlled the match with 57.45% possession compared to Newcastle United Jets FC's 42.55%.

Territorially, Auckland FC controlled 57.34% of the pitch, while Newcastle United Jets FC controlled 42.66%.

Both teams produced a similar physical output.
Auckland FC covered 97290.69 km, while Newcastle United Jets FC covered 97804.17 km.

Sprint totals were 184 for Auckland FC and 178 for Newcastle United Jets FC.

The fastest player was J. Brimmer who reached 38.1 km/h.

Tactically, Auckland FC maintained a slightly wider shape (37.46 m vs 34.86 m)
and a longer team structure (24.15 m vs 22.14 m).

The average Convex Hull Area was 595.3 m² for Auckland FC and 513.9 m² for Newcastle United Jets FC, indicating that Auckland FC occupied more space throughout the match.

Both teams maintained very similar compactness (0.640 vs 0.634), suggesting comparable defensive organization.

Overall, the tracking data sugg